ls1intum/Artemis
# `Programming exercises`: Add a grace period for course exercises

### Checklist
#### General
#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
After the due date, the repositories of programming exercises get locked to allow no more pushes by students after the deadline. For large courses, this operation can take a few seconds to finish. This leads to a small period of time where students are still allowed to submit new pushes, but their results will later be considered not graded.

This is particularly an issue for programming exercises with hidden tests. These hidden tests will always get executed on the last state of the repository. In cases like the above, this means that the latest graded result does not contain any hidden tests, while the result with hidden test result is not considered as graded:

![grafik](https://user-images.githubusercontent.com/26540346/210150173-39541f7c-4c67-4373-8926-bcbba48a4e22.png)

### Description
This PR adds a small grace period to rated results. A result is still rated if the submission date is a few seconds after the due date of the exercise.

If the grace period is exceeded, the submission is considered illegal and the result is not rated.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise

1. Wait until the due date of the exercise is over.
2. As an instructor, open the repository permission settings. Verify that the repository got locked. To simulate a delay in the locking mechanism, add push right to the student again.
3. Push a change a student. Verify that the result is still considered rated. Test that all related assessment steps (running hidden tests, manual assessment, complaints, etc) works correctly.
4. The second student pushes after the grace period (60 seconds) is over. Verify that the submission is considered illegal and the result is non-rated.

### Review Progress

#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [Result.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5085/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/Result.html) | 95% | ✅ |
| [AssessmentService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5085/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/AssessmentService.html) | 90% | ✅ |
| [ProgrammingSubmissionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5085/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingSubmissionService.html) | 94% | ✅ |
| [ProgrammingExerciseScheduleService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5085/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.scheduled/ProgrammingExerciseScheduleService.html) | 86% | ✅ |
